pkg/core/dao/dao.go

```diff
@@ -766,16 +766,20 @@ func (dao *Simple) StoreAsBlock(block *block.Block, aer1 *state.AppExecResult, a
 
 // DeleteBlock removes the block from dao. It's not atomic, so make sure you're
 // using private MemCached instance here.
-func (dao *Simple) DeleteBlock(h util.Uint256) error {
+func (dao *Simple) DeleteBlock(h util.Uint256, dropHeader bool) error {
 	key := dao.makeExecutableKey(h)
 
 	b, err := dao.getBlock(key)
 	if err != nil {
 		return err
 	}
-	err = dao.storeHeader(key, &b.Header)
-	if err != nil {
-		return err
+	if !dropHeader {
+		err = dao.storeHeader(key, &b.Header)
+		if err != nil {
+			return err
+		}
+	} else {
+		dao.Store.Delete(key)
```

Comment:
So we still do keep the full list of header hashes, right?

Reply:
Yeah, this needs additional thought and this is just 32 bytes per header (1M hashes is just boring 32M), 20 times less than ~700 bytes of the regular header.